src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs

```diff
@@ -22,6 +23,9 @@ public static class ConcurrencyUtilities
                 throw new ArgumentNullException(nameof(filePath));
             }
 
+            // limit the number of unauthorized, this should be around 30 seconds.
+            var unauthorizedAttemptsLeft = 3000;
```

Comment:
It's kind of odd with how System.IO works, but this exception is expected if another process has the lock and cleans up the file at the same time anther process is trying to get it. That means that the lock could be held for an hour (IOException) and then at the end it hits the access exception, in which case going by time would make it throw incorrectly.